src/iter_index.rs

```diff
@@ -0,0 +1,120 @@
+use core::ops::{ Range, RangeTo, RangeFrom, RangeFull, RangeInclusive, RangeToInclusive };
+use core::iter::{Skip, Take};
+use crate::Itertools;
+
+mod private_iter_index {
+    use core::ops;
+
+    pub trait Sealed {}
+
+    impl Sealed for ops::Range<usize> {}
+    impl Sealed for ops::RangeInclusive<usize> {}
+    impl Sealed for ops::RangeTo<usize> {}
+    impl Sealed for ops::RangeToInclusive<usize> {}
+    impl Sealed for ops::RangeFrom<usize> {}
+    impl Sealed for ops::RangeFull {}
+    impl Sealed for usize {}
+}
+
+/// Used by the ``range`` function to know which iterator
+/// to turn different ranges into.
+pub trait IteratorIndex<I> : private_iter_index::Sealed 
+    where I: Iterator
+{
+    /// The type that [`get`] or [`Itertools::get`]
+    /// returns when called with this type of index.
+    type Output: Iterator<Item = I::Item>;
+
+    /// Returns an iterator(or value) in the specified range.
+    ///
+    /// Prefer calling [`get`] or [`Itertools::get`] instead
+    /// of calling this directly.
+    fn index(self, from: I) -> Self::Output;
+}
+
+impl<I> IteratorIndex<I> for Range<usize>
+    where I: Iterator
+{
+    type Output = Take<Skip<I>>;
+
+    fn index(self, iter: I) -> Self::Output {
+        iter.skip(self.start)
+            .take(self.end.saturating_sub(self.start))
+    }
+}
+
+impl<I> IteratorIndex<I> for RangeInclusive<usize>
+    where I: Iterator
+{
+    type Output = Take<Skip<I>>;
+
+    fn index(self, iter: I) -> Self::Output {
+        iter.skip(*self.start())
```

Comment:
Beware, `start()` and `end()` are not specified if the `RangeInclusive` is already exhausted.